### PR TITLE
add model converter for udp

### PR DESCRIPTION
**Pull request by superyyrrzz**

### CODE REVIEW

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/SyntaxDetailViewModel.cs [hunk 1]
```diff
@@ -36,14 +36,15 @@ public class SyntaxDetailViewModel
...
         /// syntax's returns
         /// multiple return type is allowed in languages like JavaScript, Python
+        /// ApiParameter supports multiple types
         /// </summary>
-        [YamlMember(Alias = Constants.PropertyName.Returns)]
-        [JsonProperty(Constants.PropertyName.Returns)]
-        public List<ApiParameter> Returns { get; set; }
+        [YamlMember(Alias = Constants.PropertyName.Return)]
+        [JsonProperty(Constants.PropertyName.Return)]
+        public ApiParameter Return { get; set; }
```
Comment (hellosnow) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/SyntaxDetailViewModel.cs]: >ApiParameter [](start = 15, length = 12)

question: why from list to none list? how to handle multiple types

File: test/Microsoft.DocAsCode.Build.UniversalReference.Tests/UniversalReferenceDocumentProcessorTest.cs [hunk 2]
```diff
@@ -55,9 +57,52 @@ public void ProcessPythonReferencesShouldSucceed()
             {
                 var outputRawModelPath = GetRawModelFilePath(fileName);
                 Assert.True(File.Exists(outputRawModelPath));
+                var rawModel = File.ReadAllText(outputRawModelPath);
```
Comment (hellosnow) [inline test/Microsoft.DocAsCode.Build.UniversalReference.Tests/UniversalReferenceDocumentProcessorTest.cs]: >rawModel [](start = 20, length = 8)

any action to verify the raw model?

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/SyntaxDetailViewModel.cs (hunk 1 above)
Comment (superyyrrzz) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/SyntaxDetailViewModel.cs]: ApiParameter's type is a list

---
In reply to: [114244536](https://github.com/dotnet/docfx/pull/1618#discussion_r114244536) [](ancestors = 114244536)

File: test/Microsoft.DocAsCode.Build.UniversalReference.Tests/UniversalReferenceDocumentProcessorTest.cs (hunk 2 above)
Comment (superyyrrzz) [inline test/Microsoft.DocAsCode.Build.UniversalReference.Tests/UniversalReferenceDocumentProcessorTest.cs]: put in the following UT. should remove this line...

---
In reply to: [114244755](https://github.com/dotnet/docfx/pull/1618#discussion_r114244755) [](ancestors = 114244755)

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ApiHrefLinkInfoBuildOutputUrlResolver.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+{
+    using System.Web;
+
+    using AutoMapper;
+
+    public class ApiHrefLinkInfoBuildOutputUrlResolver : IValueResolver<LinkInfo, ApiLinkInfoBuildOutput, string>
+    {
+        public string Resolve(LinkInfo source, ApiLinkInfoBuildOutput destination, string destMember, ResolutionContext context)
+        {
+            var href = $"<span><a href=\"{source.LinkId}\">";
```
Comment (vwxyzh) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ApiHrefLinkInfoBuildOutputUrlResolver.cs]: >source.LinkId [](start = 42, length = 13)

html encode

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ApiNamesSpecResolver.cs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            // If href exists, return name with href
+            return GetXref(svm.Uid, svm.Name, svm.FullName);
+        }
+
+        private static string GetXref(string uid, string text = null, string alt = null)
+        {
+            var result = $"<xref href=\"{HttpUtility.HtmlEncode(uid)}\"";
+            if (!string.IsNullOrEmpty(text))
+            {
+                result += $" text=\"{text}\"";
```
Comment (vwxyzh) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ApiNamesSpecResolver.cs]: >text [](start = 37, length = 4)

html encode

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ApiNamesSpecResolver.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            }
+            else
+            {
+                result += " displayProperty=\"name\"";
+            }
+            if (!string.IsNullOrEmpty(alt))
+            {
+                result += $" alt=\"{alt}\"";
```
Comment (vwxyzh) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ApiNamesSpecResolver.cs]: >alt [](start = 36, length = 3)

html encode

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ApiNamesProfile.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+                .ForMember(dest => dest.FullName, opt => opt.Ignore())
+                .AfterMap((src, dest) =>
+                {
+                    dest.Name = ModelConverter.ToApiListInDevLangs(src.Name, src.NameInDevLangs, supportedLanguages);
+                    dest.NameWithType = ModelConverter.ToApiListInDevLangs(src.NameWithType, src.NameWithTypeInDevLangs, supportedLanguages);
+                    dest.FullName = ModelConverter.ToApiListInDevLangs(src.FullName, src.FullNameInDevLangs, supportedLanguages);
+                });
+        }
+
+    }
```
Comment (hellosnow) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ApiNamesProfile.cs]: minor: remove the empty line above

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ApiBuildOutputProfile.cs [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+                .ForMember(dest => dest.NameWithType, opt => opt.Ignore())
+                .ForMember(dest => dest.FullName, opt => opt.Ignore())
+                .ForMember(dest => dest.Source, opt => opt.Ignore())
+                .ForMember(dest => dest.AssemblyNameList, opt => opt.Ignore())
+                .ForMember(dest => dest.Platform, opt => opt.Ignore())
+                .ForMember(dest => dest.DerivedClasses, opt => opt.Ignore())
+                .ForMember(dest => dest.Implements, opt => opt.Ignore())
+                .ForMember(dest => dest.InheritedMembers, opt => opt.Ignore())
+                .ForMember(dest => dest.ExtensionMethods, opt => opt.Ignore())
+                .ForMember(dest => dest.DerivedClasses, opt => opt.Ignore())
```
Comment (hellosnow) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ApiBuildOutputProfile.cs]: >DerivedClasses [](start = 40, length = 14)

duplicate with line 27?

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ApiBuildOutputProfile.cs [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+                {
+                    opt.Condition(src => src.LinkType == LinkType.CRef);
+                    opt.MapFrom(src => src.LinkId);
+                })
+                .ForMember(dest => dest.Url, opt =>
+                {
+                    opt.Condition(src => src.LinkType == LinkType.HRef);
+                    opt.ResolveUsing(new ApiHrefLinkInfoBuildOutputUrlResolver());
+                });
+            CreateMap<ApiParameter, ApiParameterBuildOutput>();
```
Comment (hellosnow) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ApiBuildOutputProfile.cs]: >CreateMap<ApiParameter, ApiParameterBuildOutput>(); [](start = 12, length = 51)

question: need it be removed? Line 106 seems have handling ApiParameter

File: test/Microsoft.DocAsCode.Build.UniversalReference.Tests/UniversalReferenceDocumentProcessorTest.cs [hunk 9]
```diff
@@ -55,9 +57,52 @@ public void ProcessPythonReferencesShouldSucceed()
...
+            Assert.Equal("Class", model.Type);
+
+            Assert.Equal("Value", model.Name[0].Value);
+            Assert.Equal("cntk.core.Value", model.FullName[0].Value);
+
+            Assert.Equal("https://github.com/Microsoft/CNTK", model.Source[0].Value.Remote.RemoteRepositoryUrl);
+            Assert.Equal("cntk/core.py", model.Source[0].Value.Remote.RelativePath);
+            Assert.Equal(182, model.Source[0].Value.StartLine);
+
+            Assert.Equal(6, model.Syntax.Parameters.Count);
```
Comment (hellosnow) [inline test/Microsoft.DocAsCode.Build.UniversalReference.Tests/UniversalReferenceDocumentProcessorTest.cs]: >Syntax [](start = 34, length = 6)

add 1 step to verify return type?

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ModelConverter.cs [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+        {
+            return new ApiInheritanceTreeBuildOutput
+            {
+                Type = ResolveApiNames(tree.Type, supportedLanguages, references),
+                Inheritance = tree.Inheritance.Select(item => ResolveInheritanceTree(item, supportedLanguages, references)).ToList(),
+                Metadata = tree.Metadata
+            };
+        }
+
+        private static ApiBuildOutput ResolveApiBuildOutput(string uid, string[] supportedLanguages, IReadOnlyDictionary<string, ApiBuildOutput> children)
```
Comment (hellosnow) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ModelConverter.cs]: >supportedLanguages [](start = 81, length = 18)

not used in this method?

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ModelConverter.cs [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+                var message = $"Can't find {uid} in items or references";
+                Logger.LogError(message);
+                throw new DocumentException(message);
+            }
+            return result;
+        }
+
+        private static string GetXref(string uid, string text = null, string alt = null)
+        {
+            var result = $"<xref uid=\"{HttpUtility.HtmlEncode(uid)}\"";
```
Comment (hellosnow) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ModelConverter.cs]: >uid [](start = 33, length = 3)

question: why here add 'uid' to xref but ApiNamesSpecResolver add 'href'?

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ModelConverter.cs [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+                }).ToList();
+        }
+
+        public static List<ApiLanguageValuePair<ApiParameterBuildOutput>> ToApiListInDevLangsResolvingApiNames(ApiParameter defaultValue, SortedList<string, ApiParameter> values, string[] supportedLanguages, IReadOnlyDictionary<string, ApiNames> references)
+        {
+            if (defaultValue == null || supportedLanguages == null || supportedLanguages.Length == 0 || references == null)
+            {
+                return null;
+            }
+            return (List<ApiLanguageValuePair<ApiParameterBuildOutput>>)ToApiListInDevLangs(defaultValue, values, supportedLanguages)
```
Comment (hellosnow) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ModelConverter.cs]: >ToApiListInDevLangs [](start = 72, length = 19)

is it possible to write as ToApiListInDevLangs<ApiParameterBuildOutput>?

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ModelConverter.cs (hunk 11 above)
Comment (superyyrrzz) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ModelConverter.cs]: should be uid. remove the one in ApiNamesSpecResolver and use this one.

---
In reply to: [114263999](https://github.com/dotnet/docfx/pull/1618#discussion_r114263999) [](ancestors = 114263999)

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ApiBuildOutputProfile.cs (hunk 7 above)
Comment (superyyrrzz) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ApiBuildOutputProfile.cs]: remove line 27. line 50 is needed to convert derived classs.

---
In reply to: [114262767](https://github.com/dotnet/docfx/pull/1618#discussion_r114262767) [](ancestors = 114262767)

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ModelConverter.cs (hunk 12 above)
Comment (superyyrrzz) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ModelConverter.cs]: can remove (List<ApiLanguageValuePair<ApiParameterBuildOutput>>)

---
In reply to: [114265699](https://github.com/dotnet/docfx/pull/1618#discussion_r114265699) [](ancestors = 114265699)

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ApiBuildOutputProfile.cs [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+                    dest.Overridden = ModelConverter.ToApiListInDevLangsResolvingApiNames(src.Overridden, src.OverriddenInDevLangs, supportedLanguages, references);
+
+                    dest.Name = ModelConverter.ToApiListInDevLangs(src.Name, src.Names, supportedLanguages);
+                    dest.NameWithType = ModelConverter.ToApiListInDevLangs(src.NameWithType, src.NamesWithType, supportedLanguages);
+                    dest.FullName = ModelConverter.ToApiListInDevLangs(src.FullName, src.FullNames, supportedLanguages);
+                    dest.Source = ModelConverter.ToApiListInDevLangs(src.Source, src.SourceInDevLangs, supportedLanguages);
+                    dest.AssemblyNameList = ModelConverter.ToApiListInDevLangs(src.AssemblyNameList, src.AssemblyNameListInDevLangs, supportedLanguages);
+                    dest.Platform = ModelConverter.ToApiListInDevLangs(src.Platform, src.PlatformInDevLangs, supportedLanguages);
+
+                    dest.DerivedClasses = ModelConverter.ToApiListInDevLangsResolvingApiNames(src.DerivedClasses, src.DerivedClassesInDevLangs, supportedLanguages, references);
```
Comment (hellosnow) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ApiBuildOutputProfile.cs]: >DerivedClasses [](start = 25, length = 14)

also remove?

**Comment by hellosnow** [comment]
:shipit:

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ModelConverter.cs [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+            }
+            return ToApiListInDevLangs(defaultValue, values, supportedLanguages)
+                ?.Select(pair =>
+                {
+                    return new ApiLanguageValuePair<ApiNames>
+                    {
+                        Language = pair.Language,
+                        Value = ResolveApiNames(pair.Value, supportedLanguages, references)
+                    };
+                }).ToList();
```
Comment (vwxyzh) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ModelConverter.cs]: >. [](start = 18, length = 1)

?.

**Comment by vwxyzh** [comment]
:shipit:

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ModelConverter.cs (hunk 14 above)
Comment (superyyrrzz) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/ModelConverter/ModelConverter.cs]: `?.Select` in L104, It cannot be null after Select()?

---
In reply to: [114462769](https://github.com/dotnet/docfx/pull/1618#discussion_r114462769) [](ancestors = 114462769)

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildOutputs/ApiBuildOutput.cs [hunk 15]
```diff
@@ -25,11 +25,11 @@ public class ApiBuildOutput
 
         [YamlMember(Alias = Constants.PropertyName.Parent)]
         [JsonProperty(Constants.PropertyName.Parent)]
-        public ApiListInDevlangs<ApiNames> Parent { get; set; }
+        public List<ApiLanguageValuePair<ApiNames>> Parent { get; set; }
```
Comment (vicancy) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildOutputs/ApiBuildOutput.cs]: >List<ApiLanguageValuePair [](start = 15, length = 25)

why not using ApiListInDevlangs?

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildOutputs/ApiSyntaxBuildOutput.cs [hunk 16]
```diff
@@ -27,9 +27,9 @@ public class ApiSyntaxBuildOutput
         [JsonProperty("typeParameters")]
         public List<ApiParameterBuildOutput> TypeParameters { get; set; }
```
Comment (vicancy) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildOutputs/ApiSyntaxBuildOutput.cs]: can these differ per language?

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildOutputs/ApiBuildOutput.cs (hunk 15 above)
Comment (superyyrrzz) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildOutputs/ApiBuildOutput.cs]: List<ApiLanguageValuePair<T> can't be cast to ApiListInDevlangs<T> see: [http://stackoverflow.com/questions/3748931/inheriting-listt-to-implement-collections-a-bad-idea/3748942](http://stackoverflow.com/questions/3748931/inheriting-listt-to-implement-collections-a-bad-idea/3748942)

---
In reply to: [114463310](https://github.com/dotnet/docfx/pull/1618#discussion_r114463310) [](ancestors = 114463310)

**Comment by vicancy** [comment]
:shipit:

**Pull request merged**